src/build/arg/mod.rs

```diff
@@ -4851,104 +4851,106 @@ impl<'help> PartialEq for Arg<'help> {
+            write(
+                &arg.val_names
+                    .iter()
+                    .map(|n| format!("<{}>", n))
+                    .collect::<Vec<_>>()
+                    .join(&delim.to_string()),
+                true,
+            )?;
         } else {
-            " "
-        };
+            write(&format!("<{}>", arg.name), true)?;
+            if matches!(arg.num_vals, Some(x) if x > 1)
+                || (matches!(arg.num_vals, None) && (mult || mult_occ))
+            {
+                write("...", true)?;
+            }
+        }
+    } else {
+        // value of flag argument
+        write(&format!("<{}>", arg.name), true)?;
+        if mult {
+            write("...", true)?;
+        }
+    }
+    Ok(())
+}
 
+impl<'help> Display for Arg<'help> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         // Write the name such --long or -l
         if let Some(l) = self.long {
-            write!(f, "--{}{}", l, sep)?;
-        } else {
-            write!(f, "-{}{}", self.short.unwrap(), sep)?;
+            write!(f, "--{}", l)?;
+        } else if let Some(s) = self.short {
+            write!(f, "-{}", s)?;
         }
-
-        let delim = if self.is_set(ArgSettings::RequireDelimiter) {
-            self.val_delim.expect(INTERNAL_ERROR_MSG)
-        } else {
-            ' '
-        };
-
-        // Write the values such as <name1> <name2>
-        if !self.val_names.is_empty() {
-            let num = self.val_names.len();
-            let mut it = self.val_names.iter().peekable();
-
-            while let Some(val) = it.next() {
-                write!(f, "<{}>", val)?;
-                if it.peek().is_some() {
-                    write!(f, "{}", delim)?;
-                }
-            }
-
-            if self.is_set(ArgSettings::MultipleValues) && num == 1 {
-                write!(f, "...")?;
-            }
-        } else if let Some(num) = self.num_vals {
-            let mut it = (0..num).peekable();
-
-            while let Some(_) = it.next() {
-                write!(f, "<{}>", self.name)?;
-                if it.peek().is_some() {
-                    write!(f, "{}", delim)?;
-                }
-            }
-
-            if self.is_set(ArgSettings::MultipleValues) && num == 1 {
-                write!(f, "...")?;
-            }
-        } else {
-            write!(
-                f,
-                "<{}>{}",
-                self.name,
-                if self.is_set(ArgSettings::MultipleValues) {
-                    "..."
-                } else {
-                    ""
-                }
-            )?;
+        if !self.is_positional() && self.is_set(ArgSettings::TakesValue) {
+            let sep = if self.is_set(ArgSettings::RequireEquals) {
+                "="
+            } else {
+                " "
+            };
+            write!(f, "{}", sep)?;
+        }
+        if self.is_set(ArgSettings::TakesValue) || self.is_positional() {
```

Comment:
Positionals now always `TakesValue`, so this can be simplified

Reply:
This is `or` rather than `and`. I guess it can't simplified.

Reply:
Too late, but I think it can.  If positionals always have `TakesValue` set then you just need to check `TakesValue`.

```
positional = true
takes_value = true
if takes_value || positional = true
if takes_value = true

positional = false
takes_value = true
if takes_value || positional = true
if takes_value = true

positional = false
takes_value = false
if takes_value || positional = false
if takes_value = false

positional = true
takes_value = false
# Impossible
```

Reply:
XD sorry.

Reply:
*shrug* we all overlook things like that.